sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -366,6 +363,16 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
             // ((c || ...) && (d || ...)) || a || b
             (common :+ And(ldiff.reduce(Or), rdiff.reduce(Or))).reduce(Or)
           }
+        } else {
+          // No common factors from disjunctive predicates, reduce common factor from conjunction
+          val all = splitConjunctivePredicates(left) ++ splitConjunctivePredicates(right)
```

Comment:
so the improvement is to handle the mix of `&` and `|` ?

Reply:
Currently, `and` of common disjunctive predicates is optimized  Example: `(a || b || c || ...) && (a || b) => (a || b)`
but `and` of common conjunctive predicates is not optimized. Example: `((a && b) && (a && c))) => a && b && c`
The improvement is to optimize for common conjunctive predicate which can have mix of `&` and `|`

Similarly for 'Or' case of common disjunctive predicate is added in this PR.

Reply:
It seems that we are improving to handle conjunctive and disjunctive cases separately. I am wondering could we do a better job to handle the expression with mix of `&&` and `||`?

e.g.

```
(a || b) && (a && b)
```

to

```
a && b
```

Reply:
Many thanks @c21 for suggestion and review.
I was thinking about this. We can handle this as described in 3.
For `And(left, right)`
1. CommonFactorsInDisjunctions
2. else CommonFactorsInConjunctions.
3. else Mix of && and ||'s.
	```
split lhs on || to get leftOrs.
	if( leftOrs.size > 1)
		split rhs on && to get rightAnds.
		foreach p in leftOrs
			distinct = Set(p, rightAnds)
			if (distinct < rightAnds.size + 1)
				OutputOrs +=  distinct.reduce(And)
		OutputOrs.reduce(||)
	else
		split rhs on || to gt rightOrs
		if( rightOrs.size > 1)
			split lhs on && to get leftAnds.
			foreach p in rightOrs
				distinct = Set(p, leftAnds)
				if (distinct < rightAnds.size + 1)
					OutputOrs +=  distinct.reduce(And)
			OutputOrs.reduce(||)                                      //OutputOrs is a set.

```
Given example would be optimized:
```
=> (a || b) && (a && b)
=> (a && (a && b)) || (b && (a && b))
=> a && b || a && b
=> a && b
```
But for examples like this: 
```
=> (a || b) && (a && c)
=> (a && (a && c)) || (b && (a && c))
=> (a && c) || (a && b && c)
```
`(a && c) || (a && b && c)` will be reduced in next iteration of `BooleanSimplification` rule to `a && b && c`.
Hence handling mix of &&'s and ||'s in this way can be inefficient. Do you have any suggestions around how to optimize this in single iteration of `BooleanSimplification` rule.

Reply:
Thank you @Swinky ! let me think more about this and get back to you later.